### PR TITLE
Feature/vote receipt is signed by the server 199

### PR DESCRIPTION
**Pull request by colorimeter**
TCs added:
- CastVote has a signature
- CastVote signature can be verified by public key

CI changed: javabuild calls script geneate_keystore. As an effort count it as 2 TCs.

### CODE REVIEW

**Comment by magwas** [comment]
I don't believe you have to put the keystore to /etc. You can put it to target, or src/main/resources (in the latter case gitignore it).

**Comment by emddom** [comment]
SonarCloud reports bug: have you considered putting back the 
**throw .... catch**    
in MessageSigner.java

**Comment by colorimeter** [comment]
> SonarCloud reports bug: have you considered putting back the
> **throw .... catch**
> in MessageSigner.java

Why? I don't get it...

**Alive Mutant: A method call has been removed without being detected by a test. Mutation: removed call to java/security/KeyStore::load**

The reported mutant is about omitting keyStore.load, but how could have not been detected? Missing keys make both 2 TCs fail.

File: .gitignore [hunk 1]
```diff
@@ -22,3 +22,4 @@ inputs/
 issuesReport.json
 shippable/
 zentaworkaround
+src/main/resources
```
Comment (magwas) [inline .gitignore]: This shadows the whole directory out, which is probably not a good idea
I mentioned the resources directory to make it possible to lad the file through classloader, and that should have been test resources anyway.

File: src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+import javax.naming.NamingException;
+
+import org.rulez.demokracia.pdengine.exception.ReportedException;
+
+public class MessageSigner {
+
+	private PublicKey pubkey;
+	private PrivateKey privkey;
+
+    private static class Storage {
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java]: Why do we need this class?

File: src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+	        sig.update(msg);
+	        byte[] signatureBytes = sig.sign();
+	        signature = Base64.getEncoder().encodeToString(signatureBytes);
+		} catch (NoSuchAlgorithmException | InvalidKeyException | SignatureException e) {
+			throw (ReportedException)new ReportedException("Cannot compute signature").initCause(e);
+		}
+		return signature;
+	}
+
+	public static boolean verifyMessage( final byte[] msg , final String signature) {
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java]: I believe this belongs to the test code

File: src/test/java/org/rulez/demokracia/pdengine/VoteCastReceiptTest.java [hunk 4]
```diff
@@ -11,8 +12,27 @@
 
 @TestedFeature("Vote")
 @TestedOperation("Cast vote")
-@TestedBehaviour("The vote receipt contains the ballot cast and the cast vote identifier")
 public class VoteCastReceiptTest extends CreatedDefaultChoice {
+/*
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/VoteCastReceiptTest.java]: no comments, please

File: src/test/java/org/rulez/demokracia/pdengine/VoteCastReceiptTest.java [hunk 5]
```diff
@@ -23,18 +43,37 @@ public void setUp() {
 		vote.parameters.canVote = true;
 		vote.parameters.canUpdate = true;
 		vote.votesCast.clear();
+
+	  /*  MockitoAnnotations.initMocks(this);
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/VoteCastReceiptTest.java]: no comment

File: src/test/java/org/rulez/demokracia/pdengine/testhelpers/InitialContextFactoryMock.java [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+import java.util.Hashtable;
+
+import javax.naming.Context;
+import javax.naming.InitialContext;
+import javax.naming.NamingException;
+import javax.naming.spi.InitialContextFactory;
+
+public  class InitialContextFactoryMock implements InitialContextFactory {
+
+  private final static String KEYSTORE="/pdengine/src/main/resources/keystore.pk12";
```
Comment (magwas) [inline src/test/java/org/rulez/demokracia/pdengine/testhelpers/InitialContextFactoryMock.java]: do not use static path. it works without "/pdengine/".

**Review by magwas — CHANGES_REQUESTED** [review]

**Comment by magwas** [comment]
I also don't get how PIT concluded that the code works without that line. It is clearly a false positive.
A couple of sonar findings are related to that Storage class. If I could understand its intent, I could perhaps propose ways to avoid the findings.

File: src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java (hunk 2 above)
Comment (colorimeter) [reply src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java]: Using static initialization is a thread-safe singleton implementation.

File: .gitignore (hunk 1 above)
Comment (colorimeter) [reply .gitignore]: OK.

File: src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java (hunk 3 above)
Comment (colorimeter) [reply src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java]: I think the verification is a useful service, the opposite of creating the signature. I added to the class only for that reason, even if it is not yet needed anywhere. If you think it's bloatcode here, I'll remove and reimplement in the test.

File: src/test/java/org/rulez/demokracia/pdengine/VoteCastReceiptTest.java (hunk 4 above)
Comment (colorimeter) [reply src/test/java/org/rulez/demokracia/pdengine/VoteCastReceiptTest.java]: left there, sry

File: src/test/java/org/rulez/demokracia/pdengine/testhelpers/InitialContextFactoryMock.java (hunk 6 above)
Comment (colorimeter) [reply src/test/java/org/rulez/demokracia/pdengine/testhelpers/InitialContextFactoryMock.java]: OK, fine

Comment (colorimeter) [reply src/test/java/org/rulez/demokracia/pdengine/testhelpers/InitialContextFactoryMock.java]: Actually not. Changed to relative path and then the local docker image build is successful but on shippable it fails.

File: src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java (hunk 3 above)
Comment (magwas) [reply src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java]: Move it to the test code. No verification service is planned.

File: src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java (hunk 2 above)
Comment (magwas) [reply src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java]: Refer to the fields directly from within the class, make them private, make all methods non-static, and implement getInstance() as the only static method.

File: src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+public class MessageSigner {
+
+	public PublicKey pubkey;
+	private PrivateKey privkey;
+
+    private static class Storage {
+        private static final MessageSigner INSTANCE = new MessageSigner();
+     }
+
+	private MessageSigner () {
```
Comment (magwas) [inline src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java:42]: sonar is right. factor it out to more methods.

File: src/test/java/org/rulez/demokracia/pdengine/testhelpers/InitialContextFactoryMock.java (hunk 6 above)
Comment (magwas) [reply src/test/java/org/rulez/demokracia/pdengine/testhelpers/InitialContextFactoryMock.java]: I also have no idea why it fails on shippable. It cannot go to develop this way nevertheless.
Find the right way to reference it. Use the classloader for example.
And put it to the test resources, not the production ones, sorry about that.

**Review by magwas — CHANGES_REQUESTED** [review]

**Comment by magwas** [comment]
This pull request **introduces 1 alert** when merging 98aa975604419d30551ef7eb42f625587b1e7adf into 9151256dd250f10e5a0e186feaa58dffa6c35be7 - [view on LGTM.com](https://lgtm.com/projects/g/edemo/PDEngine/rev/pr-9cd304c3d339a380918878ebf9fd4575f5e2025b)

**new alerts:**

* 1 for Double\-checked locking is not thread\-safe

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: tools/generate_keystore [hunk 8]
```diff
@@ -0,0 +1,18 @@
+#!/bin/bash
+set -x
```
Comment (magwas) [inline tools/generate_keystore]: make it set  -xe to fail fast

**Comment by magwas** [comment]
This pull request **introduces 1 alert** when merging 4dc83ea6d14ddb02d4a9bfcc77d0c5efa1aa636c into 9151256dd250f10e5a0e186feaa58dffa6c35be7 - [view on LGTM.com](https://lgtm.com/projects/g/edemo/PDEngine/rev/pr-9a887ea447b88e60503473a2f27285b8abb083a1)

**new alerts:**

* 1 for Double\-checked locking is not thread\-safe

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by magwas** [comment]
This pull request **introduces 1 alert** when merging 2894096366b0415950f12d87ab9e93f2437cbee0 into 9151256dd250f10e5a0e186feaa58dffa6c35be7 - [view on LGTM.com](https://lgtm.com/projects/g/edemo/PDEngine/rev/pr-b6b47805ce84f11cac5ff5d27bf8cd8cfa5118e4)

**new alerts:**

* 1 for Double\-checked locking is not thread\-safe

---

*Comment posted by [LGTM.com](https://lgtm.com)*

File: src/test/java/org/rulez/demokracia/pdengine/testhelpers/InitialContextFactoryMock.java (hunk 6 above)
Comment (colorimeter) [reply src/test/java/org/rulez/demokracia/pdengine/testhelpers/InitialContextFactoryMock.java]: OK, put to target/test-classes/resources, it works on shippable too (/pdengine bind mount is not available in shippable container).

File: src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java (hunk 7 above)
Comment (colorimeter) [reply src/main/java/org/rulez/demokracia/pdengine/MessageSigner.java:42]: separated to three parts.

File: tools/generate_keystore (hunk 8 above)
Comment (colorimeter) [reply tools/generate_keystore]: OK.

**Comment by colorimeter** [comment]
Are there any more issues yet to address?

File: Makefile [hunk 9]
```diff
@@ -39,9 +39,14 @@ testenv:
 
 javabuild: target/PDEngine-0.0.1-SNAPSHOT.jar
 
-target/PDEngine-0.0.1-SNAPSHOT.jar:
+target/PDEngine-0.0.1-SNAPSHOT.jar: maven-prepare keystore maven-build
+
+maven-prepare:
 	mvn build-helper:parse-version versions:set versions:commit -DnewVersion=\$${parsedVersion.majorVersion}.\$${parsedVersion.minorVersion}.\$${parsedVersion.incrementalVersion}-$$(tools/getbranch|sed 'sA/A_Ag').$$(git rev-parse --short HEAD)
-	mvn clean org.jacoco:jacoco-maven-plugin:prepare-agent install org.pitest:pitest-maven:mutationCoverage site -Pintegration-test
```
Comment (magwas) [inline Makefile:44]: why?

**Review by magwas — CHANGES_REQUESTED** [review]

**Review by magwas — APPROVED** [review]

**Comment by magwas** [comment]
It seems that the project did not compile. I don't see the reason. Any thoughts?

File: tools/prepare [hunk 10]
```diff
@@ -5,7 +5,7 @@ then
     Xvnc4 -SecurityTypes none :0 &
 fi
 set +x
-. <(./tools/decrypt_secret_env $ORG_NAME)
+source ./tools/decrypt_secret_env $ORG_NAME
```
Comment (magwas) [inline tools/prepare]: this should be the culprit. Please revert this change

**Review by magwas — CHANGES_REQUESTED** [review]

**Comment by colorimeter** [comment]
My understanding of current sonar issues:
- hazards are false positives, without loading the encryption keys the test fails. (Maybe the dynamic code change in jvm of hazard testing cannot cope with the already initialized static content)
- cyclomatic complexity of the whole class is high: IMHO it is simply caused by the huge amount of exceptions which can occur. They are many branches under the hood, but for us they do not have too much meaning, since they all reported by exceptions in which the originating cause is propagated.

**Review by magwas — APPROVED** [review]

**Comment by magwas** [comment]
testcasecount=3
